src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DataMemberBindingMetadataProvider.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Reflection;
+using System.Runtime.Serialization;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Mvc.ModelBinding.Metadata
+{
+    /// <summary>
+    /// An <see cref="IBindingMetadataProvider"/> for <see cref="DataMemberAttribute.IsRequired"/>.
+    /// </summary>
+    public class DataMemberRequiredValidationMetadataProvider : IBindingMetadataProvider
+    {
+        /// <inheritdoc />
+        public void GetBindingMetadata([NotNull] BindingMetadataProviderContext context)
+        {
+            // Types cannot be required; only properties can
+            if (context.Key.MetadataKind != ModelMetadataKind.Property)
+            {
+                return;
+            }
+
+            var dataMemberAttribute = context
+                .Attributes
+                .OfType<DataMemberAttribute>()
+                .FirstOrDefault();
+            if (dataMemberAttribute == null || !dataMemberAttribute.IsRequired)
+            {
+                return;
+            }
+
+            // isDataContract == true iff the container type has at least one DataContractAttribute
+            var containerType = context.Key.ContainerType.GetTypeInfo();
+            var isDataContract = containerType.GetCustomAttribute<DataContractAttribute>() != null;
+            if (isDataContract)
+            {
+                // We don't need to add a validator, just to set IsRequired = true. The validation
+                // system will do the right thing.
+                context.BindingMetadata.IsRequired = true;
+            }
```

Comment:
This thing used to be a validator provider and it would return a no-op 'required' validator. Just setting `IsRequired = true` does the same thing. So I decided to do that instead.